airbyte-ci/connectors/base_images/base_images/python/v1.py

```diff
@@ -0,0 +1,79 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+"""This module declares all the airbyte python connector base image for version 1.
+Please create a v2.py module if you want to declare a new major version.
+"""
+from __future__ import annotations
+
+from typing import Final, final
+
+import dagger
+from base_images import common, sanity_checks
+from base_images.python.common import AirbytePythonConnectorBaseImage, PythonBase
+
+
+class _1_0_0(AirbytePythonConnectorBaseImage):
```

Comment:
These base images in theory make it easy to roll out improvements to all our connectors without having to
1. Update 300 dockerfiles
2. Overrite an existing docker image

Meaning we can rollout features and fixes instantly

and If we follow semver where
1. patch = fix
2. minor = feature
3. major = breaking functionality

Would we ever want to reference anything but major versions in our codebase?

If we want to maintain version history of patch and minor for third party devs that seems to be a concern when we start publishing actual images to dockerhub. And with that we get minor and patch version history for free

Reply:
By breaking functionality are we aligned that we mean that connector change is required to use the new base image version?

I think we still have scenario in which we cut non-breaking versions
* Feature: a perf improvement, let's 
* Patch: a patch update of Python (Python 3.9.19?) 

If we consider feature/patch are not worth a new image version it means we'll overwrite a dockerhub tag, and it means we lose reproductible builds: even if the image tag does not change the connector will get built from a different image.

Reply:
> We always want to release patch and minor updates to the base image to all python connectors

This maybe where there's a misunderstanding. I don't want a connector to be silently updated to a new base image version without a metadata change. Mainly because of build reproducibility. In the design I'm implementing the actual base version upgrade happens with a connector PR. A new version declaration has no effect per se if no connector is using it. The plan is to make an automation that'll autobump the connector base image version on each new connector PR. We'll have connectors referencing multiple base image versions in the wild. If we want to batch upgrade all our connectors we'll have to open a big PR modifying their metadata (this can be scripted / automated) of course.

Reply:
Ah ok! I like that were coming up to design goals. Means that the decision is easier (A or B) kind of thing.

I believe there are two areas to talk about, let me break them out

**1. Reference by major or by patch**

So my interpretation of this project was through the lens of our Pyaml incident where we had **wanted** to patch all python connectors **but could not** as the patch would have to happen in the middle of all there dockerfile.

I was hoping that this daggerization of these base images would make an issue like this go away for us.

And I think that is in line with the [tech spec](https://docs.google.com/document/d/19XTGfqsHFnEc_LHlalUwERLw6n8JDYOnflDdPiex0OM/edit) we have

![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/PTsI7qAmiIMkhFQg04QF/a1bd0ab4-6b78-436f-9a0c-9d54fc3b7d3e/image.png)

Given the above it seems to imply we would want connectors to reference a major version and receive security patches for free

**2. What happens when a library or a utility is updated?**

Before I used the example of the dagger library being updated breaking an old version, and I dont think weve discussed, what happens then if all versions are both in code and in the master branch?

I also what to add the example of what happens when we update a utility function like `with_python_module`?

That in theory effects all base images? correct? If so theres a few issues I see
1. Do we want to always be cutting patches for small refactors?
2. Doesnt the update of utilities fundamentally alter past base versions? For example imagine we use `with_python_module` in all base images, we are on latest v1.2.3, we change the implementation of `with_python_module`, does that not technically change all connectors includeing something like 1.0.0? And as a result violate "build reproducibility"?


Anyway let me know if we need to schedule a call about this and thanks for being so open to the larger critique. Im glad that we both get to work towards whatever the implementation becomes :)

Reply:
**1. Reference by major or by patch**
I really don't want silent upgrade of connector image, even if it's a patch version. If we cut a patch new version with bug we'll have no rollout control and it can lead to some strange situation. 
To roll out upgrade at scale I believe the commands I introduced in https://github.com/airbytehq/airbyte/pull/30520 are a good starting point. So, to rephrase the tech spec:
Instead of editing different dockerfiles we'll: 
- Cut a new patch version
- Run an `airbyte-ci connectors upgrade-base-image` command
- Open and merge PR that will publish the updated connectors.
With this approach we keep scale but also tracability thanks to the git history.

**2. What happens when a library or a utility is updated?**
I understand this concern. What I'm attached to is as way for the library to expose the existing versions. Would it sound good to you to:
* Always keep a single class for the current base image version
* At build step we: publish the current version to dockerhub, persist a versionned registry in YAML/JSON of the available image. This registry is exposed as a global variable on the package.
* The `baseImage` metadata will then consume a "real" image from dockerhub, but will be told which one to use from the registry in this lib.

Reply:
Thanks for taking the feedback. Im definitely liking the direction of this and could be on board. Just some clarifying questions

A few questions resulting from us going back to a docker image (from dagger) instead of a dagger image
1. Do we loose any important dagger functionality? e.g. cache volumes, things we might want to parameterize like poetry version, or injecting custom logic in the middle of build logic?
2. Does this change where/how you configure the baseImage to use or even the post install hook? Does a dockerfile then make more sense. (I really dont want to reopen this debate, just felt it should be asked so I can understand the `airbyte-ci build` implementation differences between a airbyte python connector and a third party connect with a dockerfile)

A question resulting from us now using the dockerhub registry
1. Do we still need a VersionRegistry and/or our own yaml/json file defining what docker images are available? doesnt `docker ls` or `crane ls` fill that need?

Reply:
> A few questions resulting from us going back to a docker image (from dagger) instead of a dagger image
1. No we don't lose anything. If our connector build logic continues to interact with the `VersionRegistry` it will continue to receive dagger container objects. The single difference is that the VersionRegistry will build container `from_("dockerhub:image")`. 
2. This does not change has to change anything to what I've implemented so far in https://github.com/airbytehq/airbyte/pull/30456 (or the pre/post build hook) as the connector build pipeline will receive the same object: a Container. 

> Do we still need a VersionRegistry and/or our own yaml/json file defining what docker images are available? doesnt docker ls or crane ls fill that need?

I like the `VersionRegistry` being a proxy between our connector build pipeline and our remote registry because:
- It will map "internal" version names to actual registry images. If we change our registry (let's say we migrate from DockerHub to ghcr.io) we can just change the pointer in the `VersionRegistry` and the new remote registry will be used.
- If we ever want to deprecate an image we can remove it from the `VersionRegistry`. It will not break third party that rely on it because the image will still be hosted, I will just break our internal build system. 
- We keep control over what a `previous` and `latest` base image version mean, which can simplify some roll out scenarios. 

I believe this `VersionRegistry` debate to be an implementation detail at this step. I find it convenient for the build pipeline to consume a python object directly without `docker` or `crane` call. But if we ever want to decouple the current package from `pipelines` we can quite easily change it and make the connector build instantiate the base container object from the remote registry with the `from_` method.